src/core/server/elasticsearch/client/retry_unauthorized.ts

```diff
@@ -0,0 +1,134 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { MaybePromise } from '@kbn/utility-types';
+import { AuthHeaders, KibanaRequest, SetAuthHeaders, isRealRequest } from '../../http';
+import { ScopeableRequest } from '../types';
+import { UnauthorizedError } from './errors';
+
+/**
+ * @public
+ */
+export interface UnauthorizedErrorHandlerOptions {
+  error: UnauthorizedError;
+  request: KibanaRequest;
+}
+
+/**
+ * @public
+ */
+export interface UnauthorizedErrorHandlerResultRetryParams {
+  authHeaders: AuthHeaders;
+}
+
+/**
+ * @public
+ */
+export interface UnauthorizedErrorHandlerRetryResult
+  extends UnauthorizedErrorHandlerResultRetryParams {
+  type: 'retry';
+}
+
+/**
+ * @public
+ */
+export interface UnauthorizedErrorHandlerNotHandledResult {
+  type: 'notHandled';
+}
+
+/**
+ * @public
+ */
+export type UnauthorizedErrorHandlerResult =
+  | UnauthorizedErrorHandlerRetryResult
+  | UnauthorizedErrorHandlerNotHandledResult;
+
+/**
+ * Toolkit passed to a {@link UnauthorizedErrorHandler} used to generate responses from the handler
+ * @public
+ */
+export interface UnauthorizedErrorHandlerToolkit {
+  /**
+   * The handler cannot handle the error, or was not able to reauthenticate
+   * */
+  notHandled: () => UnauthorizedErrorHandlerNotHandledResult;
+  /**
+   * The handle was able to reauthenticate.
+   * @param params
+   */
+  retry: (params: UnauthorizedErrorHandlerResultRetryParams) => UnauthorizedErrorHandlerRetryResult;
+}
+
+/**
+ * A handler used to handle unauthorized error returned by elasticsearch
+ *
+ * @public
+ */
+export type UnauthorizedErrorHandler = (
+  options: UnauthorizedErrorHandlerOptions,
+  toolkit: UnauthorizedErrorHandlerToolkit
+) => MaybePromise<UnauthorizedErrorHandlerResult>;
+
+/** @internal */
+export type InternalUnauthorizedErrorHandler = (
+  error: UnauthorizedError
+) => MaybePromise<UnauthorizedErrorHandlerResult>;
+
+/** @internal */
+export const toolkit: UnauthorizedErrorHandlerToolkit = {
+  notHandled: () => ({ type: 'notHandled' }),
+  retry: ({ authHeaders }) => ({
+    type: 'retry',
+    authHeaders,
+  }),
+};
+
+const notHandledInternalErrorHandler: InternalUnauthorizedErrorHandler = () => toolkit.notHandled();
+
+/**
+ * Converts the public version of `UnauthorizedErrorHandler` to the internal one used by the ES client
+ *
+ * @internal
+ */
+export const createInternalErrorHandler = ({
+  handler,
+  request,
+  setAuthHeaders,
+}: {
+  handler: UnauthorizedErrorHandler;
+  request: ScopeableRequest;
+  setAuthHeaders: SetAuthHeaders;
+}): InternalUnauthorizedErrorHandler => {
+  // we don't want to support 401 retry for fake requests
+  if (!isRealRequest(request)) {
+    return notHandledInternalErrorHandler;
+  }
+  return async (error) => {
+    try {
+      const result = await handler({ request, error }, toolkit);
```

Comment:
__question:__ I see that right now the request that is being passed to `handler` contains auth headers (`Authorization: xxx`) that were set by the Core and weren't present in the original request. Would it be possible to "unset" these headers before passing request object to the `handler`? 

Right now Security basically skips such requests since it assumes that it was the client who specified authentication headers explicitly. We can try to workaround this in Security plugin and unset these headers on our own, but I'm wondering if you have a better way of passing original "pristine" request object instead?

Reply:
Did you progress on your investigations?

As you said, the mutation is done here:

https://github.com/elastic/kibana/blob/8ffc8728e684d28275bbd2130506be8e11b29cbc/src/core/server/http/http_server.ts#L447-L452

But I don't see any way to differentiate headers that have been added by this snippet from other headers, so apart from just removing this part of the code (which would likely break some stuffs relying on these headers to forge fake requests, right?) I'm not sure how we could answer your ask?

Reply:
> Did you progress on your investigations?

I can think of only one clumsy hack on Security side: https://github.com/elastic/kibana/pull/122155/files#diff-df4d732995701a873f86257177d6440fa562a820a0c849624cc1e54de2bbfbf1R238-R244 (`Object.fromEntries(Object.entries(...))` is needed because I cannot just remove `authorization` header from the `headers` object since it's `readonly`).
```ts
(request.headers as Record<string, unknown>) = Object.fromEntries(
  Object.entries(request.headers).filter(
    ([headerName]) => headerName.toLowerCase() !== 'authorization'
  )
);
```

> so apart from just removing this part of the code (which would likely break some stuffs relying on these headers to forge fake requests, right?

Yeah, reporting (and probably Upgrade Assistant in some cases) still depends on this I believe.

> But I don't see any way to differentiate headers that have been added by this snippet from other headers

I was thinking that you might have a copy of the auth headers returned during authentication stage so that you know which headers are original and which were added as a part of the authentication flow, or it's not the case?

Reply:
> I was thinking that you might have a copy of the auth headers returned during authentication stage so that you know which headers are original and which were added as a part of the authentication flow, or it's not the case?

Nope, not really.

Well, we do set the headers returned from the auth hook inside the `authRequestHeaders` map, so in a way, we can identify which entries in `req.headers` where injected by lookings at the keys of the associated entry in `authRequestHeaders`, but:

- This will be quite ugly if we do that from the ES error interception logic. We're already leaking responsibilities all over the place, and it will just increase it.
- In case of overrides (e.g a header returned from the login hook overriding a header initially present on the request), we won't be able to provide the initial value of such headers.
- `KibanaRequest` is immutable, and we won't be able to easily recreate one to provide it to the hook

Reply:
> This will be quite ugly if we do that from the ES error interception logic. We're already leaking responsibilities all over the place, and it will just increase it.

Yeah, I see. It feels like there is no way to avoid an ugly workaround of some sort, it's just a matter where it will live. I'm fine with keeping it in Security if there is no elegant way to accomplish this directly in the Core.

> In case of overrides (e.g a header returned from the login hook overriding a header initially present on the request), we won't be able to provide the initial value of such headers.

It's something that should never happen, at least Security will never override existing headers.

> KibanaRequest is immutable, and we won't be able to easily recreate one to provide it to the hook

If we agree to keep workaround (e.g. rewriting `request.headers`), is it safe to assume that `request.headers` will stay writable?

Reply:
> is it safe to assume that request.headers will stay writable?